ENH: Add pyflowy

HaoZeke opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [X] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [X] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

amritagos commented:
> @amritagos @MSallermann can you confirm that you'll be co-maintainers for the `conda-forge` recipe?

Sure, I'll confirm! Thanks! :)

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyflowy/meta.yaml```) and found it was in an excellent condition.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyflowy/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipes/pyflowy/recipe.yaml**:

 * ❌ ``requirements: build: xtensor < 0.25.0`` should not contain a space between relational operator and the version, i.e. ``xtensor <0.25.0``

For **recipes/pyflowy/recipe.yaml**:

 * ℹ️ Jinja2 variable references are suggested to take a ``${{<one space><variable name><one space>}}`` form. See lines [38].


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16695590029](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16695590029). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyflowy/recipe.yaml```) and found it was in an excellent condition.

HaoZeke commented:
Same error as https://github.com/conda-forge/staged-recipes/pull/30703#issuecomment-3146602767

```bash
1 warning and 20 errors generated.
 │ │   [6/9] Compiling C++ object subprojects/flowy/libflowy.dylib.p/src_topography.cpp.o
 │ │   FAILED: [code=1] subprojects/flowy/libflowy.dylib.p/src_topography.cpp.o
 │ │   x86_64-apple-darwin13.4.0-clang++ -Isubprojects/flowy/libflowy.dylib.p -Isubprojects/flowy -I../subprojects/flowy -I../subprojects/flowy/thirdparty/tsl -Isubprojects/pdf_cpplib -I../subprojects/pdf_cpplib -I$PREFIX/include -I$BUILD_PREFIX/include -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -Wall -Winvalid-pch -Wextra -Wpedantic -std=c++20 -O3 -g -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/pyflowy-0.0.1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -mmacosx-version-min=10.13 -mmacosx-version-min=10.13 -Wno-unused-local-typedefs -Wno-array-bounds -ffast-math -fno-finite-math-only -fno-plt -Bsymbolic -DWITH_NETCDF -MD -MQ subprojects/flowy/libflowy.dylib.p/src_topography.cpp.o -MF subprojects/flowy/libflowy.dylib.p/src_topography.cpp.o.d -o subprojects/flowy/libflowy.dylib.p/src_topography.cpp.o -c ../subprojects/flowy/src/topography.cpp
 │ │   In file included from ../subprojects/flowy/src/topography.cpp:3:
 │ │   In file included from ../subprojects/flowy/flowy/include/topography.hpp:8:
 │ │   $BUILD_PREFIX/include/xtensor/xsort.hpp:297:27: warning: unused parameter 'data_end' [-Wunused-parameter]
 │ │     297 |             ConstRandomIt data_end,
 │ │         |                           ^
 │ │   In file included from ../subprojects/flowy/src/topography.cpp:4:
 │ │   In file included from ../subprojects/flowy/flowy/include/asc_file.hpp:4:
 │ │   ../subprojects/flowy/flowy/include/topography_file.hpp:33:65: error: 'path' is unavailable: introduced in macOS 10.15 unknown - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
 │ │      33 |     std::filesystem::path handle_suffix( const std::filesystem::path & path_in )
 │ │         |                                                              
```

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory. You can also examine the [workflow logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16698068187) for more detail.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16698068187](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16698068187). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pyflowy/recipe.yaml```) and found it was in an excellent condition.

synapticarbors commented on recipes/pyflowy/LICENSE:
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
You should remove this file since it is included in the source distribution and conda/rattler will pick it up from there.

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+tests:
+  - python:
+      imports:
+        - pyflowy
```
you can remove the `script` section below and just add
`pip_check: true`. See:
https://rattler.build/latest/reference/recipe_file/#python-tests

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+context:
+  name: pyflowy
+  version: 0.0.1
+  build_num: 0
```
Instead of using `context` to set jinja variables for `build_num` and `shasum`, just set them explicitly below.

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+
+build:
+  skip:
+    - if: python < 3.9
```
Technically you don't need to have the `python < 3.9` check since conda-forge only builds packages for python >=3.9

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+    - win
+    - osx
+  number: ${{ build_num }}
+  string: h${{ hash }}_${{ build_num }}
```
Don't set the build string. That's created automatically.

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+    - pybind11 >=2.12.0
+    - libnetcdf
+    - hdf5
+    - ${{ pin_compatible('numpy') }}
```
You can just change this to `numpy`. `pin_compatible is no longer needed:
https://conda-forge.org/docs/maintainer/knowledge_base/#building-against-numpy

I like to still include `numpy` explicitly in `run`, even though there is a `run_exports` from the numpy package.

synapticarbors commented on recipes/pyflowy/variants.yaml:
```diff
@@ -0,0 +1,5 @@
+c_stdlib_version:
```
Unless there is a good reason to specify this, it is best to just use the conda-forge defaults. My suggestion would be to delete this file.

synapticarbors commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,88 @@
+    - numpy
+    - libnetcdf
+    - hdf5
+    - ${{ compiler('c') }}
```
`compiler(...)` should go up in the `build` section. I believe `host` and `build` should be reorganized. There's several things that should be in the `build_platform != target_platform` section. The following recipe might be a good guide:
https://github.com/conda-forge/matplotlib-feedstock/blob/5da242b3aa49613299e2bc176011d22aac266bf3/recipe/meta.yaml

HaoZeke commented:
@conda-forge/help-python-c, ready for review!

(bumping as a reminder)

carterbox commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,85 @@
+    - hdf5
+    - libnetcdf
+    - numpy
+    - pybind11 >=2.12.0
```
Why do you need pybind11 at runtime?

danielnachun commented on recipes/pyflowy/recipe.yaml:
```diff
@@ -0,0 +1,85 @@
+  run:
+    - fmt
+    - hdf5
+    - libnetcdf
```
```suggestion
```
These libraries should already have run exports, so they shouldn't need to be explicitly added here.

HaoZeke committed:
ENH: Add pyflowy recipe

Adds conda-forge recipe for pyflowy (Python bindings for flowy).
Addresses review feedback: fix solver conflict by dropping xtensor
pin, use pkg-config, pin pybind11 >=2.12.0 for numpy 2 compat,
skip Windows, bump macOS deployment target to 10.15 for
std::filesystem support.

HaoZeke commented:
@conda-forge/help-python-c, ready for review!

HaoZeke committed:
Address review: remove win skip, delete conda_build_config.yaml

- Remove win platform skip per danielnachun (enable all platforms, fix what fails)
- Delete conda_build_config.yaml per xhochy (macOS default is now 11.0)

HaoZeke commented:
@conda-forge/help-python-c, ready for review!

HaoZeke committed:
fix(pyflowy): remove unnecessary sed, enable macOS builds

Remove the sed line that stripped pybind11 from pyproject.toml
runtime deps -- with --no-deps --no-build-isolation, pip does not
install runtime dependencies anyway.

Re-enable macOS builds (skip only win). The earlier macOS failure
was std::filesystem::path unavailable on 10.13 SDK; with the
conda-forge default now at 11.0, this should be resolved.

Addresses review comments from traversaro and danielnachun.

HaoZeke committed:
Merge branch 'main' into addPyFlowy

HaoZeke committed:
fix(pyflowy): drop incorrect pybind11 runtime Requires-Dist for pip check

Build and import already succeeded; tests failed because upstream PyPI
metadata lists pybind11 as install_requires while only numpy is needed at
runtime. Strip that line from the installed METADATA after pip install.

HaoZeke committed:
fix(pyflowy): locate METADATA via SP_DIR for post-install patch

sysconfig purelib pointed at BUILD_PREFIX during the staged-recipes
build, so the pybind11 Requires-Dist strip never found the installed
files. Use conda-forge SP_DIR (with PREFIX fallback).

HaoZeke commented:
@conda-forge/help-python-c, ready for review!

CI green on linux/osx (win intentionally skipped) + linter. Fixed the failing `pip check` by stripping upstream's incorrect runtime `Requires-Dist: pybind11` from the installed METADATA (pybind11 is build-only; run deps are numpy + python).

traversaro commented:
> > Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
> 
> Can you comment a bit on this? Is this package vendoring other packages (as it seems the case from the licenses) or not? What is the rationale for vendoring these packages instead of having as separate conda-forge packages?

Can you please comment on this @HaoZeke ?

HaoZeke commented:
@traversaro Yes — the PyPI sdist vendors three trees, and the recipe packages every corresponding license:

1. **flowy** (`subprojects/flowy`, GPL-3.0) — the C++ core from https://github.com/flowy-code/flowy. Built as a **static** library into the `pyflowy.flowycpp` extension (`default_library=static`); we do not install a separate `libflowy` from this recipe.
2. **pdf_cpplib** (`subprojects/pdf_cpplib`, GPL-3.0) — small internal PDF helpers used by flowy (https://github.com/amritagos/pdf_cpplib). Not on conda-forge today.
3. **tsl robin_map/set** (`subprojects/flowy/thirdparty/tsl`, MIT) — header-only, vendored by upstream flowy. Note `tsl_robin_map` already exists on conda-forge.

External deps that *are* on conda-forge are not vendored: `fmt`, `hdf5`/`libnetcdf`, `xtensor` / `xtensor-blas` / `xtensor-python`, `numpy`, `pybind11`.

**Rationale for the current shape:** upstream’s install path is a single sdist that pulls flowy (and its wraps) via meson so `pip install pyflowy` works without a preinstalled C++ stack. There is no `flowy` / `pdf_cpplib` feedstock yet, so the bindings cannot depend on them as run/host packages.

**Longer term:** splitting `flowy` (and optionally `pdf_cpplib`) into their own feedstocks and switching this recipe to consume those plus `tsl_robin_map`, then dropping the subproject trees. Working on those staged-recipes PRs next; can rework this PR to depend on them once they land, or keep the vendored path if you prefer merge first.

HaoZeke commented:
Follow-up staged-recipes PR for the unbundled C++ pieces: https://github.com/conda-forge/staged-recipes/pull/34125 (`pdf_cpplib` + `flowy`, system `tomlplusplus` / `tsl_robin_map` / `xtensor`). Once those feedstocks exist, this pyflowy recipe can drop the vendored subprojects and depend on them.
